libs/dkg/examples/beacon_manager/main.cpp

```diff
@@ -0,0 +1,149 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "crypto/bls_base.hpp"
+#include "crypto/bls_dkg.hpp"
+#include "dkg/beacon_manager.hpp"
+
+#include <cstdint>
+#include <iostream>
+#include <random>
+#include <stdexcept>
+#include <unordered_map>
+#include <vector>
+
+using namespace fetch::crypto;
+using namespace fetch::dkg;
+using namespace fetch::byte_array;
+
+int main(int argc, char **argv)
+{
+  if (argc < 2)
+  {
+    std::cerr << "usage: " << argv[0] << " [cabinet size]" << std::endl;
+    return -1;
+  }
+
+  // Initializing the BLS library
+  bls::Init();
+
+  // Beacon parameters
+  std::unordered_map<Identity, std::shared_ptr<BeaconManager>> nodes;
+  uint64_t cabinet_size = static_cast<uint64_t>(atoi(argv[1]));
+  uint32_t threshold    = static_cast<uint32_t>(cabinet_size >> 1);
+
+  // Creating nodes
+  for (uint64_t i = 0; i < cabinet_size; ++i)
+  {
+    std::shared_ptr<BeaconManager> node = std::make_shared<BeaconManager>();
+    node->Reset(cabinet_size, threshold);
+    nodes[node->identity()] = node;
+  }
+
+  // Communicating BLS identities
+  using ParticipantDetails = std::pair<Identity, bls::Id>;
+  std::vector<ParticipantDetails> participants;
+  for (auto &m : nodes)
+  {
+    auto &n = *m.second;
+    participants.push_back({n.identity(), n.id()});
+  }
+
+  // Propagating identities
+  for (auto &n : nodes)
+  {
+    // Shuffling to similate random arrival order
+    std::random_shuffle(participants.begin(), participants.end());
```

Comment:
There is an issue using random_shuffle  without that breaks mac builds - it happens to @baykaner compiler. Do this instead:

```
    std::random_device rd;
    std::mt19937       g(rd());
    std::shuffle(participants.begin(), participants.end(), g);

```